bugsnag-android-core/src/main/java/com/bugsnag/android/BugsnagMainJournal.kt

```diff
@@ -0,0 +1,175 @@
+                    break
+                }
+                Thread.sleep(highWaterPollingIntervalMS)
+            }
+        }
+    }
+
+    companion object {
+        // Top-level key where version info will be saved.
+        internal const val versionInfoKey = "version-info"
+        internal const val journalTypeKey = "type"
+        internal const val journalVersionKey = "version"
+
+        // The "type" of the main journal. This will probably never change.
+        internal const val journalType = "Bugsnag Android"
+
+        // Version of this journal. This will change for migration purposes.
+        internal const val journalVersion = 1
+
+        // Size of the on-disk memory-mapped buffer in bytes.
+        private const val mmapBufferSize = 100000L
+
+        // If the memory-mapped buffer fills beyond this many bytes, it gets auto-snapshotted.
+        // Note: The size of mmapBufferSize and highWater have no effect on snapshot runtime cost.
+        private const val highWater = 50000L
+
+        // Polling interval in milliseconds for the high water check thread.
+        private const val highWaterPollingIntervalMS = 200L
+
+        /**
+         * The global instance of the main journal.
+         * You must call startInstance() before accessing this.
+         */
+        val instance: BugsnagMainJournal
+            get() = mainJournal ?: throw IllegalStateException("Must call startInstance() before accessing instance")
+
+        /**
+         * Start the main journal at the specified base path.
+         * Any existing journal at this location will be overwritten.
+         *
+         * NOTE: This must be called once (and only once) before accessing instance.
+         *
+         * @param logger A logger
+         * @param baseDocumentPath Path to a filename that will be used as the basis for all journal and snapshot files
+         * @param initialDocument The initial contents of the document. This will be deep copied.
+         * @return The contents of the previous journal (if any)
+         */
+        fun startInstance(
+            logger: Logger,
+            baseDocumentPath: File,
+            initialDocument: Map<String, Any>
+        ): MutableMap<in String, out Any>? {
+            synchronized(hasInitialized) {
+                if (hasInitialized) {
+                    return null
+                }
+
+                val previousContents = try {
+                    loadPrevious(baseDocumentPath)
+                } catch (exc: IOException) {
+                    null
+                }
+
+                val journal = BugsnagMainJournal(logger, baseDocumentPath, initialDocument)
+                mainJournal = journal
+                hasInitialized = true
+                journal.beginSnapshotHousekeepingThread()
+                return previousContents
+            }
+        }
+
+        internal var hasInitialized = false
+        internal var mainJournal: BugsnagMainJournal? = null
```

Comment:
This looks like it might want to be `lateinit`:
```suggestion
        internal lateinit var mainJournal: BugsnagMainJournal
```